### PR TITLE
190 Add package change confirmation

### LINKED ISSUE #190: Add package change confirmation to publish CLI command
With the completion of #127 we will be removing packages that are no longer present in the package.yaml.  We should add confirmation of what functions will be added and removed when the publish command is run.  For example:

```shell
$ functionary package publish ../examples/demo/
Publishing package "demo".

Environment
FunctionFoundary - dev

New
- somethingnew

Removed
- output_text

Updated
- long_running_process
- output_json

Continue [y|n]? 
```

Each section could show only if there is data to be shown there.  Also, we could have an optional flag `-y` that could be provided to bypass the confirmation and just publish.

The above is just for demonstration purposes and not intended to be strict guidance on the formatting.

### PR DESCRIPTION
**Pull request by lights7412**
Closes #190

Checks if there are changes between the package.yaml and the api when publishing. If there are changes it will display the associated changes and ask if you want to continue.

The confirmation can be skipped by adding a -y to the publish command <functionary package publish ~/functionary/examples/demo -y>.

Three Unit tests were added in the cli/tests/tests_utils.py. Further testing can be done by adjusting the package.yaml (ie adding, deleting, changing functions)

### CODE REVIEW

File: cli/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import pytest
+
+from functionary.utils import new_functions, removed_functions, updated_functions
+
+
+def function_creator(num):
+    function_list = {}
+    for i in range(num):
+        function_dict = {}
+        function_dict["name"] = "Test Name" + str(i)
```
Comment (scott-taubman) [inline cli/tests/test_utils.py:10]: use f-strings:

```python
function_dict["name"] = "Test Name{i}"
```

**Review by scott-taubman — CHANGES_REQUESTED** [review]

File: cli/functionary/package.py [hunk 2]
```diff
@@ -101,15 +101,30 @@ def create_cmd(ctx, language, name, output_directory):
...
+
+    Use the -y flag to bypass the confirmation of changes
+    and publish.
     """
+    if os.path.exists(path + "/package.yaml"):
+        changes = check_changes(path + "/package.yaml")
+        if not y and changes:
+            confirm = input("Continue [y|N]? ").lower()
+            if confirm not in ("y", "yes"):
+                exit(1)
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/package.py:126]: fixed

File: cli/functionary/utils.py [hunk 3]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
         table.add_row(*row_data)
         first_row = False
     console.print(table)
+
+
+def package_functions(path):
+    """Returns list of functions in the package.yaml"""
+    in_file = open(path, "rb")
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:99]: fixed

File: cli/functionary/utils.py [hunk 4]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+    package_contents_io = in_file.read()
+    in_file.close()
+    package_definition = yaml.safe_load(package_contents_io)
+    return package_definition["package"]
+
+
+def api_functions(packageName):
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:106]: fixed

File: cli/functionary/package.py [hunk 5]
```diff
@@ -101,15 +101,30 @@ def create_cmd(ctx, language, name, output_directory):
...
     Use the -k option to keep the build artifacts
     (found in $HOME/.functionary/builds) after publishing,
     rather than cleaning it up.
+
+    Use the -y flag to bypass the confirmation of changes
+    and publish.
     """
+    if os.path.exists(path + "/package.yaml"):
+        changes = check_changes(path + "/package.yaml")
+        if not y and changes:
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/package.py:123]: fixed

File: cli/functionary/utils.py [hunk 6]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+    return updatedfunctions
+
+
+def bullet_list(list, title):
+    """
+    Prints to console
+    Title in italics and list in bullet form
+    """
+    print("\x1B[3m" + title + "\x1B[0m")
+    for i in range(len(list)):
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:234]: fixed

File: cli/functionary/utils.py [hunk 7]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+            updatedfunctions[packagefunctions[i].get("name")] = changedfield
+    return updatedfunctions
+
+
+def bullet_list(list, title):
+    """
+    Prints to console
+    Title in italics and list in bullet form
+    """
+    print("\x1B[3m" + title + "\x1B[0m")
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:233]: fixed

File: cli/functionary/utils.py [hunk 8]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+    for i in range(len(list)):
+        print("   ~", list[i].get("name"), "\n")
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:235]: fixed

File: cli/functionary/utils.py [hunk 9]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+                if packagefunctions[i].get("return_type") != apifunctions[j].get(
+                    "return_type"
+                ):
+                    changedfield.append("Return Type")
+        if changedfield:
+            updatedfunctions[packagefunctions[i].get("name")] = changedfield
+    return updatedfunctions
+
+
+def bullet_list(list, title):
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:228]: fixed, is generic

File: cli/functionary/utils.py [hunk 10]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+    """
+    print("\x1B[3m" + title + "\x1B[0m")
+    for i in range(len(list)):
+        print("   ~", list[i].get("name"), "\n")
+
+
+def format_updated_functions(results, title=""):
+    """
+    Print The function name and changed fields
+    in a table
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:241]: fixed and added an editor ruler to my settings

File: cli/tests/test_utils.py (hunk 1 above)
Comment (lights7412) [reply cli/tests/test_utils.py:10]: fixed

File: cli/functionary/package.py [hunk 11]
```diff
@@ -101,15 +101,30 @@ def create_cmd(ctx, language, name, output_directory):
...
     Publish a package to make it available in the currently active environment.
 
     Use the -k option to keep the build artifacts
     (found in $HOME/.functionary/builds) after publishing,
     rather than cleaning it up.
+
+    Use the -y flag to bypass the confirmation of changes
+    and publish.
     """
+    if os.path.exists(path + "/package.yaml"):
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/package.py:121]: fixed

File: cli/functionary/utils.py [hunk 12]
```diff
@@ -89,3 +92,161 @@ def format_results(results, title="", excluded_fields=[]):
...
+    for function in functions:
+        package_id = function["package"]
+        function_dict = {}
+        function_dict["name"] = function["name"]
+        function_dict["display_name"] = function["display_name"]
+        function_dict["summary"] = function["summary"]
+        function_dict["description"] = function["description"]
+        function_dict["return_type"] = function["return_type"]
+
+        if package_id in functions_lookup:
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py:136]: fixed. Also adjusted package list command to call the sort_functions_by_package function

File: cli/functionary/utils.py [hunk 13]
```diff
@@ -89,3 +93,150 @@ def format_results(results, title="", excluded_fields=[]):
...
+        updatedfunctions = updated_functions(packagefunctions, apifunctions)
+
+    if newfunctions:
+        bullet_list(newfunctions, "New functions")
+    if removedfunctions:
+        bullet_list(removedfunctions, "Removed Functions")
+    if updatedfunctions:
+        _format_updated_functions(updatedfunctions, "Updated Functions")
+
+    return newfunctions or removedfunctions or updatedfunctions
```
Comment (scott-taubman) [inline cli/functionary/utils.py:168]: There's a little bit of ambiguity with what you are returning here.  `newfunctions or removedfunctions or updatedfunctions` will return the first of those values that isn't empty.  You then check that value for truthiness later on.  What would be more explicit would be to return a boolean here.

**Review by scott-taubman — CHANGES_REQUESTED** [review]
Overall this looks much improved and my remaining comments are pretty minor.

One issue that I failed to test for before is that the check raises an exception when publishing a new package because it assumes it's going to find a package in the API call.  You'll need to handle the case where there's no existing package to compare the one being published to.

File: cli/functionary/utils.py (hunk 13 above)
Comment (lights7412) [reply cli/functionary/utils.py]: fixed

File: cli/functionary/utils.py [hunk 14]
```diff
@@ -89,3 +93,150 @@ def format_results(results, title="", excluded_fields=[]):
...
+
+def check_changes(path):
+    """
+    Checks for discrepancies between the package.yaml and the API functions, returns
+    whether or not changes occurred
+    """
+    packagefunctions = _get_package_functions(path)["functions"]
+    apifunctions = _get_functions_for_package(_get_package_functions(path)["name"])
+
+    # Nothing in the api
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py]: removed

File: cli/functionary/utils.py [hunk 15]
```diff
@@ -89,3 +93,150 @@ def format_results(results, title="", excluded_fields=[]):
...
+        if package["name"] == package_name:
+            return package["id"]
+
+
+def sort_functions_by_package(functions):
+    """
+    Sorts functions by their package id
+    """
+    functions_lookup = {}
+    package_id = 0
```
[the remark this answers is not in the document]
Comment (lights7412) [reply cli/functionary/utils.py]: removed

**Pull request merged**